docs/proposals/cloud-provider-refactoring.md

```diff
@@ -0,0 +1,121 @@
+
+### 3. Strategy for refactoring Kubelet
+
+The majority of the calls by the kubelet to the cloud is done during the initialization of the Node Object. The other uses are for configuring Routes (in case of GCE), scrubbing DNS, and periodically polling for IP addresses.
+
+All of the above steps, except the Node initialization step can be moved into a controller. Specifically, IP address polling, and configuration of Routes can be moved into the cloud dependent controller manager.
+
+Scrubbing DNS, after discussing with @thockin, was found to be redundant. So, it can be disregarded. It is being removed.
+
+Finally, Node initialization needs to be addressed. This is the trickiest part. Pods will be scheduled even on uninitialized nodes. This can lead to scheduling pods on incompatible zones, and other wierd errors. Therefore, an approach is needed where kubelet can create a Node, but mark it as "NotReady". Then, some asynchronous process can update it and mark it as ready. This is now possible because of the concept of ExternalAdmissionControllers introduced by @justinsb in these PRs - (https://github.com/kubernetes/kubernetes/pull/36210, https://github.com/kubernetes/kubernetes/pull/36209)
+
+Admission controllers are meant to perform some aspects of initialization of objects. In this case, an ExternalAdmissionController will be configured to first mark the Node object as unusable by setting the NodeStatus as {"CloudProviderInitialized": "False"} (This exact keyword has not been decied yet), and then an external controller (residing in the cloud controller manager) will update it by removing the NodeStatus. In order for this to work, Nodes should be unusable until these Statuses are removed. This is possible because of the above mentioned PRs by @justinsb.
+
+### 4. Strategy for refactoring Kube-ApiServer
+
+Kube-apiserver uses the cloud provider for two purposes
+
+1. Distribute SSH Keys - This can be moved to a controller
+2. Admission Controller for PV - This can be refactored using the External Admission Controller approach used in Kubelet
+
+### 5. Strategy for refactoring Volumes
+
+Volumes need cloud providers, but they only need SPECIFIC cloud providers.  The real problem is that cloud providers can take params from cloud config. The majority of the logic for volume management, currently resides in the controller-manager.
+
+There is an undergoing effort to move all of the volume logic from the controller-manager into plugins called Flex Volumes. In the Flex volumes world, all of the vendor specific code will be packaged in a separate binary as a plungin. After discussing with @thockin, this was decidedly the best approach to remove all cloud provider dependency for volumes out of kubernetes core.
+
+### 6. Roadmap
+
+As the first step, an opt-in flag will be provided to run the external cloud controller manager. This is being aimed to be completed by 1.6 release.
```

Comment:
This will be the `--cloud-provider=external` flag, right?
Please point that out...